test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,373 @@
+			continue
+		}
+
+		// Skip deletion if there is still a virtual machine which refers this IP address.
+		if virtualMachineIPAddresses[ip.Spec.Address] {
+			continue
+		}
+
+		ipAddressClaimsToDelete = append(ipAddressClaimsToDelete, ipAddressClaim)
+	}
+
+	for _, ipAddressClaim := range ipAddressClaimsToDelete {
+		ipAddressClaim := ipAddressClaim
+		Byf("Deleting IP Addressclaim %s/%s", ipAddressClaim.GetNamespace(), ipAddressClaim.GetName())
+		if err := ipamClient.Delete(ctx, &ipAddressClaim); err != nil {
+			errList = append(errList, errors.Wrapf(err, "Deleting IPAddressClaim %s/%s", ipAddressClaim.GetNamespace(), ipAddressClaim.GetName()))
+		}
+	}
+
+	if len(errList) > 0 {
+		return kerrors.NewAggregate(errList)
+	}
+	return nil
+}
+
+// getVirtualMachineIPAddresses lists all VirtualMachines in the given folder and
+// returns a map which contains the IP addresses of all machines.
+// If the given folder is not found it will return an error.
+func getVirtualMachineIPAddresses(ctx context.Context, folderName string, vSphereClient *govmomi.Client) (map[string]bool, error) {
+	finder := find.NewFinder(vSphereClient.Client)
+
+	// Find the given folder.
+	folder, err := finder.FolderOrDefault(ctx, folderName)
+	if err != nil {
+		return nil, err
+	}
+
+	// List all VirtualMachines in the folder.
+	managedObjects, err := finder.ManagedObjectListChildren(ctx, folder.InventoryPath+"/...", "VirtualMachine")
+	if err != nil {
+		return nil, err
+	}
+
+	var vm mo.VirtualMachine
+	virtualMachineIPAddresses := map[string]bool{}
+
+	// Iterate over the VirtualMachines, get the `guest.net` property and extract the IP addresses.
+	for _, mobj := range managedObjects {
+		// Get guest.net properties for mobj.
+		if err := vSphereClient.RetrieveOne(ctx, mobj.Object.Reference(), []string{"guest.net"}, &vm); err != nil {
+			// if err := vmobj.Properties(ctx, vmobj.Reference(), []string{"guest.net"}, &vm); err != nil {
+			return nil, errors.Wrapf(err, "Get properties of VM %s", mobj.Object.Reference())
+		}
+		// Iterate over all nics and add IP addresses to virtualMachineIPAddresses.
+		for _, nic := range vm.Guest.Net {
+			if nic.IpConfig == nil {
+				continue
+			}
+			for _, ip := range nic.IpConfig.IpAddress {
+				virtualMachineIPAddresses[ip.IpAddress] = true
+			}
+		}
+	}
+
+	return virtualMachineIPAddresses, nil
+}
+
+type ipamService struct {
+	client          client.Client
+	ipAddressClaims []ipamv1.IPAddressClaim
+	skipCleanup     bool
+}
+
+func newIPAMService(e2eIPAMKubeconfig string, skipCleanup bool) (*ipamService, error) {
+	ipamClient, err := getClientSet(e2eIPAMKubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ipamService{
+		ipAddressClaims: []ipamv1.IPAddressClaim{},
+		client:          ipamClient,
+		skipCleanup:     skipCleanup,
+	}, nil
+}
+
+func getClientSet(e2eIPAMKubeconfig string) (client.Client, error) {
```

Comment:
Would it make sense to call this getClient?

ClientSet is usually the client-go thing which (IIRC) is actually a set of clients